### ISSUE TITLE
long-running `/search` query on matrix.org after client disconnected

### ISSUE THREAD

**Issue by anoadragon453** [matrix-org/synapse#14290]
Discovered a long-running `SearchRestServlet` (search text of all rooms) on matrix.org that was running for a user that is only in 21 rooms. But perhaps one of the rooms is large.

Logs:

```
$ grep "POST-2984873" homeserver.log
2730070:2022-10-25 07:30:34,357 - synapse.handlers.search - 142 - INFO - POST-2984873 - Search batch properties: None, None, None
2730071:2022-10-25 07:30:34,358 - synapse.handlers.search - 149 - INFO - POST-2984873 - Search content: {'search_categories': {'room_events': {'search_term': 'update', 'filter': {'limit': 10}, 'order_by': 'recent', 'event_context': {'before_limit': 1, 'after_limit': 1, 'include_profile': True}}}}
2755696:2022-10-25 07:33:34,346 - synapse.http.site - 372 - INFO - POST-2984873 - Connection from client lost before response was sent
4542688:2022-10-25 11:06:10,238 - synapse.handlers.search - 327 - INFO - POST-2984873 - Found 0 events to return
4542689:2022-10-25 11:06:10,239 - synapse.http.server - 744 - WARNING - POST-2984873 - Not sending response to request <XForwardedForRequest at 0x7f979c09ae50 method='POST' uri='/_matrix/client/r0/search' clientproto='HTTP/1.1' site='8080'>, already disconnected.
4542690:2022-10-25 11:06:10,239 - synapse.access.http.8080 - 460 - INFO - POST-2984873 - 194.126.177.75 - 8080 - {@[redacted]:matrix.org} Processed request: 12935.892sec/-12755.893sec (0.004sec, 0.001sec) (0.003sec/12935.878sec/8) 0B 200! "POST /_matrix/client/r0/search HTTP/1.1" "[redacted]" [0 dbevts]
```

See that the request was processing for ~3.5hrs. There was a spike in DB traffic at the very end of the event as well. Looks like a good endpoint candidate to make cancelable.

**Comment by squahtx** [comment]
Do we know whether the 3.5 hours was all in a single big DB query or lots of smaller queries?
Cancellation would help with the latter, but not the former.

**Comment by anoadragon453** [comment]
Looking at `SearchStore.search_msgs`, I believe the bulk is in a single query unfortunately.

https://github.com/matrix-org/synapse/blob/81815e0561eea91dbf0c29731589fac2e6f98a40/synapse/storage/databases/main/search.py#L428-L545

**Comment by DMRobertson** [comment]
If I was executing a long-running query in `psql` I could press `^C` to interrupt it. Does psycopg2 offer a similar mechanism that we could trigger on cancellation?

**Comment by squahtx** [comment]
> If I was executing a long-running query in `psql` I could press `^C` to interrupt it. Does psycopg2 offer a similar mechanism that we could trigger on cancellation?

There is, according to stackoverflow.
https://stackoverflow.com/a/2275123:

> You can cancel a query by calling the [`pg_cancel_backend(pid)`](http://www.postgresql.org/docs/8.3/static/functions-admin.html#FUNCTIONS-ADMIN-SIGNAL-TABLE) PostgreSQL function in a separate connection.
> 
> You can know the PID of the backend to cancel from the [`connection.get_backend_pid()`](http://initd.org/psycopg/docs/connection.html#connection.get_backend_pid) method of psycopg2 (available from version 2.0.8).

Plumbing everything together is the challenging part. The cancellation has to go from the main thread to the db threadpool thread running the transaction, and there's twisted code in the way.

**Comment by squahtx** [comment]
I suspect that @erikjohnston's proposed fix in #14402 has resolved this issue, so I shall close it.

We'll reopen the issue if we see an excessively slow `/search` again.
